gapis/api/vulkan/frame_loop.go

```diff
@@ -126,129 +139,237 @@ func newFrameLoop(ctx context.Context, c *capture.GraphicsCapture, numInitialCmd
+			}
+
+			// Do mid-loop stuff.
+			{
+				// Iterate through the loop contents, emitting instructions one by one.
+				for cmdIndex, cmd := range f.capturedLoopCmds {
+
+					// We've already processed the first loop command, so skip that one.
+					// All others get emitted.
+					if cmdIndex > 0 {
+						out.MutateAndWrite(ctx, f.capturedLoopCmdIds[cmdIndex], cmd)
+					}
+				}
+			}
+
+			// Do end loop stuff.
+			{
+				apiState := GetState(out.State())
+
+				stateBuilder := apiState.newStateBuilder(ctx, newTransformerOutput(out))
+				defer stateBuilder.ta.Dispose()
+
+				// Write out the command that ended the loop. That one is outside of the captured commands since it's the one that called this code.
+				out.MutateAndWrite(ctx, cmdId, cmd)
+
+				// Notify the other transforms that we have just emitted the end of the loop.
+				out.NotifyPostLoop(ctx)
+
+				// Now we need to emit the instructions to reset the state, before the conditional branch back to the start of the loop.
+				if err := f.resetResource(ctx, stateBuilder); err != nil {
+					log.E(ctx, "FrameLoop: Failed to reset changed resources %v.", err)
+					return
+				}
+
+				// Add conditional jump instruction to bring us back to the start of the loop while we've not done.
+				stateBuilder.write(stateBuilder.cb.Custom(func(ctx context.Context, s *api.GlobalState, b *builder.Builder) error {
+					b.Load(protocol.Type_Int32, f.loopCountPtr)
+					b.Sub(1)
+					b.Clone(0)
+					b.Store(f.loopCountPtr)
+					b.JumpNZ(uint32(0x1))
+					return nil
+				}))
+			}
+
+			// Finally, we've done all the processing for a loop. Nothing left to do.
+			return
+
+		} else { // We're currently inside the loop.
+
+			// Lets just remember the command we've seen so we can do all the work we need at the end of the loop.
+			// This is done because the information we need to transform the loop is only available at that time;
+			// due to the possibility of preceeding transforms modifing the loop contents in-flight.
+			f.capturedLoopCmds = append(f.capturedLoopCmds, cmd)
+			f.capturedLoopCmdIds = append(f.capturedLoopCmdIds, cmdId)
+
+			log.D(ctx, "FrameLoop: inside loop at frame %v, cmdId %v, cmd %v.", f.frameNum, cmdId, cmd)
+
 			return
 		}
 
-		// Add jump instruction
-		sb.write(sb.cb.Custom(func(ctx context.Context, s *api.GlobalState, b *builder.Builder) error {
-			b.Load(protocol.Type_Int32, f.loopCountPtr)
-			b.Sub(1)
-			b.Clone(0)
-			b.Store(f.loopCountPtr)
-			b.JumpNZ(uint32(0x1))
-			return nil
-		}))
+	} else { // We're after the loop. Again, we can simply pass-through commands.
+		out.MutateAndWrite(ctx, cmdId, cmd)
 		return
 	}
 
-	if _, ok := cmd.(*VkQueuePresentKHR); ok {
-		f.frameNum++
-	}
-
-	out.MutateAndWrite(ctx, id, cmd)
-
+	// Should have early out-ed before this point.
+	log.F(ctx, true, "FrameLoop: Internal control flow error: Should not be possible to reach this statement.")
```

Comment:
We should panic if we ever hit this case.

Reply:
done